### ISSUE TITLE
Second damaging battlerage skill (skill 4) not in use [version 1.9-pre-ae01b23-moveBattlerageToAttainment]

### ISSUE THREAD

**Issue by ose-goetia** [AchaeaBashingScript/Bashing#62]
I downloaded version 1.9-pre-ae01b23-moveBattlerageToAttainment.

Using both the 'simple' and 'simplereverse' strategies, bashing is not using my second damaging rage (skill 4) (throatrip as a serpent), and a sentinel confirmed it was only using their first battlerage attack. I am using my autoraze skill (skill 3) properly.

Portals and theft are present in my battlerage list. I am not sure how to check the value of battlerageSkills[4] to confirm if the skill is present or not.

**Comment by keneanung** [comment]
Could you give me the output of `AB ATTAINMENT`?

**Comment by ose-goetia** [comment]
Abilities in Attainment:
-------------------------------------------------------------------------------
Portals              Godly Rescue
Thrash               Use your whip to open painful, bloody wounds.
Battlerage           The rage of battle.
Flagellate           Sensitize your target's nerves.
Market               Sell, sell, sell!
Battle               Danger awaits!
Limitedportals       Growing independence.
Excoriate            Use your rage to flay a denizen's shield.
Independence         A newbie no more.
Embrace              Full membership in your class.
Tradeskills          Learn a trade.
Theft                Everyone's a target.
Throatrip            Go for the jugular.
Snare                Trap your target in a tight grip.
Craftsmanship        Learn another trade.
Obliviate            Induce forgetfulness in your target.
Multiclass           Learn an additional class.
Sustenance           Sleep is for the weak. Food too.
   -------------------------------------------------------------------------
                  *** Next ability available in 7 levels ***
   -------------------------------------------------------------------------
Dragonhood           The soul of the Greater Dragon.
Polymath             Learn another additional class.
-------------------------------------------------------------------------------
To gain further information on an ability, AB <ability>.

**Comment by adayoung** [comment]
It's because gmcp.IRE.Display.ButtonActions is broken for some classes. That makes keneanung.bashing.rageAvailable go weird. Here's what I'm seeing for Runewarden:
```
{
  button2 = {
    text = "Portals",
    highlight = 0,
    commands = "portals"
  },
  button6 = {
    text = "Market",
    highlight = 0,
    commands = [[
marketon
marketoff
market <text>]]
  },
  button1 = {
    text = "Jab",
    highlight = 0,
    commands = "JAB @tar"
  },
  button7 = {
    text = "Battle",
    highlight = 0,
    commands = ""
  },
  button5 = {
    text = "Bulwark",
    highlight = 0,
    commands = "bulwark"
  },
  button4 = {
    text = "Battlerage",
    highlight = 0,
    commands = ""
  },
  button3 = {
    text = "Collide",
    highlight = 0,
    commands = "collide @tar"
  }
}
```

**Comment by adayoung** [comment]
And here's gmcp.IRE.Display.ButtonActions for Dragon:
```
{
  button2 = {
    text = "Portals",
    highlight = 1,
    commands = "portals"
  },
  button6 = {
    text = "Market",
    highlight = 1,
    commands = [[
marketon
marketoff
market <text>]]
  },
  button1 = {
    text = "Gut",
    highlight = 0,
    commands = "GUT @tar"
  },
  button7 = {
    text = "Battle",
    highlight = 1,
    commands = ""
  },
  button5 = {
    text = "Glaciate",
    highlight = 0,
    commands = "glaciate @tar"
  },
  button4 = {
    text = "Battlerage",
    highlight = 1,
    commands = ""
  },
  button3 = {
    text = "Dragonchill",
    highlight = 0,
    commands = "dragonchill @tar"
  }
}
```

**Comment by adayoung** [comment]
Yeah. Here's what I got for blademaster:
```
{
  button2 = {
    text = "Portals",
    highlight = 1,
    commands = "portals"
  },
  button6 = {
    text = "Market",
    highlight = 1,
    commands = [[
marketon
marketoff
market <text>]]
  },
  button1 = {
    text = "Drawslash",
    highlight = 0,
    commands = "DRAWSLASH @tar"
  },
  button7 = {
    text = "Battle",
    highlight = 1,
    commands = ""
  },
  button5 = {
    text = "Daze",
    highlight = 0,
    commands = "shin daze @tar"
  },
  button4 = {
    text = "Battlerage",
    highlight = 1,
    commands = ""
  },
  button3 = {
    text = "Leapstrike",
    highlight = 0,
    commands = "leapstrike @tar"
  }
}
```

**Comment by adayoung** [comment]
Oops, I hope I'm not spamming by now :blush: Perhaps we need to use cooldowns instead of those buttons?

**Comment by adayoung** [comment]
What if we have another table to say whether a battlerage skill is on a cooldown, update it whenever a battlerage ability is used, and flip it back with a temptimer? `keneanung.bashing.rageAvailable` is essentially looking at cooldown from serverside right? :open_mouth:

**Comment by keneanung** [comment]
That and if enough rage is available for the ability. Yes, we could do tempTimers or timeframes, but I loved the idea to let the server do the work for me 😏

And I hope IRE cares enough about their nexus users to fix those buttons.

It's probably best to add at least a failsafe in case the buttons are broken (like now) or the user removes their rage buttons. This solution still comes from a time where these buttons were fixed IIRC.

**Comment by adayoung** [comment]
Hi! I opened #63 for this and with that, it's working as it used to! :grin:

**Comment by adayoung** [comment]
@Agzohmqueks, you're about the third black dragon that has complained on this. Let me have a look at it this weekend. Can I pounce you in-game to run some tests as well?!

**Comment by Agzohmqueks** [comment]
Yeah I'm happy to help @ada - I may not be able to be in the same room as you for IC reasons, but am happy to feed you my output or whatever you need so send me a message or tell. I'm pretty good with code myself as well so I can follow along with trialling the changes if you want.

**Comment by adayoung** [comment]
@Agzohmqueks, @ose-goetia can you guys try this release: https://github.com/AchaeaBashingScript/Bashing/releases/tag/v1.9-pre-1a2076b-moveBattlerageToAttainmentIREButtonActions .. and let me know how that works out for you?

**Comment by adayoung** [comment]
Wait wait, hang on.. wow IRE is bad.

**Comment by adayoung** [comment]
Right then, use this release https://github.com/AchaeaBashingScript/Bashing/releases/download/v1.9-pre-da30238-moveBattlerageToAttainmentIREButtonActions/Bashing.mpackage
@Agzohmqueks @ose-goetia you'll want to look at `kconfig bashing` and ensure it shows the correct battlerage skills for all your classes. If you click on the 'reset' button there, try skipping or running around or spamming till you see `keneanung: Finished parsing battlerage skills.` It takes some time to get the skills out of there.

**Comment by Agzohmqueks** [comment]
Ok cool. Can't wait to try it. If I hit the reset button it's not going to reset my prio list as well?

**Comment by Agzohmqueks** [comment]
Works perfectly. Even with my custom battlerage strat and bashing system. Thank you for your hard work! P.S. I'm going to get in touch about contributing to this and other github projects. I might need help getting modules setup to sync with github.

**Comment by Agzohmqueks** [comment]
Oh, one thing more quirk I've noticed though. From what I can tell, this happens: if you have enough rage for an ability but it doesn't get used within a cooldown window, because you killed the target or fled (anything where the stopAttack function is called, sorry I forget the exact spelling), but then total rage fades to zero because of overall rage cooldown, then when you startAttack again it tries to use the available rage attack as if you had rage. I suspect (haven't had a proper look at the code) this is because there is no new timer made for losing rage globally when you haven't gained any rage within the cooldown window.

Does that make sense?

**Comment by adayoung** [comment]
Total rage available is actually [read and updated](https://github.com/AchaeaBashingScript/Bashing/blob/da30238389865d701a9023b78b4a9a744d9ed22c/script.lua#L1285-L1291) whenever the event gmcp.Char.Vitals goes off, and that's pretty much each prompt. There's no internal timer tracking any overall rage cooldown as such. I've personally never seen it attempting to use a rage attack without actually having enough rage. Try using the simple strategy just to see if it's not something in your custom thing?

**Comment by Agzohmqueks** [comment]
It is quite possible then, that what I'm witnessing is either the attack being queued off balance whilst I do have rage, and firing when I've lost it, or my rage dropping to zero a few milliseconds before I attack, and the attack is being added to queue before gmcp gets updated because of my server ping, which is huge because Australia. It's also possible I don't have any check for "total rage available" in my custom strat, where the simple ones do. I'll compare the coding and see.

As it doesn't effect anything (I could gag that line and it would all still work fine), I'll keep using it and try to come up with a more definite diagnosis. Hopefully by then I've got GitHub syncing setup and can just offer my own cganges.
